### ISSUE TITLE
CP-SAT solver returning non-optimal solutions with status "OPTIMAL"

### ISSUE THREAD

**Issue by martineronneberg** [google/or-tools#3639]
**What version of OR-Tools and what language are you using?**
Version: 9.5.2237
Language: Python 3.10.8

**Which solver are you using (e.g. CP-SAT, Routing Solver, GLOP, BOP, Gurobi)**
CP-SAT

**What operating system (Linux, Windows, ...) and version?**
Windows 10

**What did you do?**
I am noticing some inconsistencies in the solutions that are returned from the CP-SAT solver, as they seem to not be optimal even though the status is optimal. Detailed description below: 

 I have a minimization optimization problem, where I have set the solver time limit to 7 minutes. In the original formulation I got this result when solving the problem: 

Problem solved in 122.55572287300001 seconds
Problem solved with 41325 search branches

CpSolverResponse summary:
status: OPTIMAL
objective: 1.103767340707094e+17
best_bound: 115986843582.3407
booleans: 12738
conflicts: 8770
branches: 41325
propagations: 2929920
integer_propagations: 7734418
restarts: 17013
lp_iterations: 56393
walltime: 122.556
usertime: 122.556
deterministic_time: 40.8764
gap_integral: 236.054
solution_fingerprint: 0x1f0ef30f9fb14810

However, it is possible to quickly manually verify that this is not in fact an optimal solution, even though the status indicates this. Therefore, I tried adding a solution printer:

`solution_printer = cp_model.ObjectiveSolutionPrinter()`
`status = solver.Solve(model, solution_printer)`

This gave the following result: 
Solution 0, time = 28.27 s, objective = 382720787664
Solution 1, time = 28.90 s, objective = 310493081856
Solution 2, time = 29.57 s, objective = 116348607376
Solution 3, time = 32.58 s, objective = 116024831952
Solution 4, time = 33.93 s, objective = 116021831952
Solution 5, time = 34.50 s, objective = 116021081952
Solution 6, time = 37.53 s, objective = 116021079088
Solution 7, time = 38.11 s, objective = 116015829088
Solution 8, time = 38.84 s, objective = 115995073344
Solution 9, time = 42.66 s, objective = 115994745936
Solution 10, time = 43.23 s, objective = 115990555472
Solution 11, time = 44.78 s, objective = 115989511344
Solution 12, time = 48.94 s, objective = 115988460800

Problem solved in 421.097292254 seconds
Problem solved with 105057 search branches

CpSolverResponse summary:
status: FEASIBLE
objective: 1.103767340707094e+17
best_bound: 115986210034.6411
booleans: 12629
conflicts: 59457
branches: 105057
propagations: 9880171
integer_propagations: 22945204
restarts: 17014
lp_iterations: 59780
walltime: 421.097
usertime: 421.097
deterministic_time: 168.686
gap_integral: 2259.42
solution_fingerprint: 0xc9b32282d27b219

Now, the status is more correct, as it switched from optimal to feasible, However, the solution objective value is the same. It is also using more time. I also notice that the objective value in solution 12 is actually better than the one in the response summary. 

I also tried to limit the solution space, by adding a constraint putting an upper limit to one of the variables in the objective, as I was seeing that this variable had value > 0 even when it did not need to in order to be feasible. This gave better results, but the objective value it is still not equally as good as last solution (solution 10). It is also easy to verify manually that this is in fact NOT an optimal solution: 

Solution 0, time = 38.72 s, objective = 382720787664
Solution 1, time = 39.46 s, objective = 310493081856
Solution 2, time = 40.11 s, objective = 116348607376
Solution 3, time = 43.75 s, objective = 116348536304
Solution 4, time = 44.15 s, objective = 116027045440
Solution 5, time = 44.78 s, objective = 116026803632
Solution 6, time = 45.81 s, objective = 116023803632
Solution 7, time = 46.76 s, objective = 116022260736
Solution 8, time = 49.06 s, objective = 115999244128
Solution 9, time = 53.80 s, objective = 115989951824
Solution 10, time = 55.27 s, objective = 115988460800

Problem solved in 348.895701323 seconds
Problem solved with 90383 search branches

CpSolverResponse summary:
status: OPTIMAL
objective: 146577823485
best_bound: 115986843582.3407
booleans: 12668
conflicts: 39989
branches: 90383
propagations: 5701151
integer_propagations: 13741912
restarts: 17026
lp_iterations: 59355
walltime: 348.896
usertime: 348.896
deterministic_time: 142.237
gap_integral: 923.263
solution_fingerprint: 0xde714f73fbddc045

**Summing up, am I misunderstanding how to read this output, or do I need to use another solver to get closer to optimal solutions? Is the solution I am getting at the end not the solution with the lowest objective value of all the solutions found?**

**Comment by lperron** [comment]
please send us the proto: model.ExportToFile('issue3639.pb.txt')

Please remove callbacks, and just add the parameter

    solver.parameters.log_search_progress = True

**Comment by martineronneberg** [comment]
[model_proto.zip](https://github.com/google/or-tools/files/10443781/model_proto.zip)

I have zipped and uploaded the model proto files, and removed callback and added the parameter as you suggested.

**Comment by lperron** [comment]
On the main branch, 

```
Initial optimization model '': (model_fingerprint: 0xbe6d0e5a0a817985)                                                
#Variables: 1116871 (#ints:283290 in floating point objective)                                                        
  - 65340 Booleans in [0,1]                                                                                           
  - 2774 different domains in [0,12283314082588] with a largest complexity of 1.                 
  - 617 constants in {0,1}                                                                                            
#kBoolOr: 292152 (#literals: 292152)                                                                                  
#kLinear1: 61979                                                                                                      
#kLinear2: 415844                                                                                                     
#kLinear3: 116089                                                                                                     
#kLinearN: 170869 (#terms: 2238220)                                                                                   
                                                                                                                      
Starting presolve at 0.38s                                                                                            
[Scaling] Floating point objective has 283230 terms with magnitude in [1, 7.12253e+06] average = 1.60452e+06          
[Scaling] Objective coefficient relative error: 1                                                                     
[Scaling] Objective worst-case absolute error: 9.80776e+16                                                            
[Scaling] Objective scaling factor: 0.0625          

...

#Model  33.72s var:5180/16238 constraints:4532/15039 compo:2664,1369,382,64,52,49,22,22,22,18,...
#Bound  39.74s best:1.15988461e+11 next:[1.15987817e+11,1.15988461e+11] pseudo_costs
#Model  39.83s var:4789/16238 constraints:4123/15039 compo:2019,1192,560,278,51,41,34,22,22,22,...
#Bound  40.37s best:1.15988461e+11 next:[1.15987817e+11,1.15988461e+11] pseudo_costs
#Bound  47.97s best:1.15988461e+11 next:[1.15987818e+11,1.15988461e+11] pseudo_costs
#Bound  49.10s best:1.15988461e+11 next:[1.15987818e+11,1.15988461e+11] pseudo_costs
#Model  52.76s var:4786/16238 constraints:4121/15039 compo:2018,1190,560,278,51,41,34,22,22,22,...
#Bound  54.35s best:1.15988461e+11 next:[1.15987819e+11,1.15988461e+11] pseudo_costs
#Bound  55.13s best:1.15988461e+11 next:[1.15987819e+11,1.15988461e+11] pseudo_costs
#Model  61.16s var:4774/16238 constraints:4109/15039 compo:2006,1190,560,278,51,41,34,22,22,22,...
#Bound  99.52s best:1.15988461e+11 next:[1.15987844e+11,1.15988461e+11] pseudo_costs
#Model  99.60s var:4750/16238 constraints:4084/15039 compo:1982,1190,560,278,51,41,34,22,22,22,...
#Bound 100.48s best:1.15988461e+11 next:[1.1598798e+11,1.15988461e+11] max_lp
#Bound 102.52s best:1.15988461e+11 next:[1.15987985e+11,1.15988461e+11] pseudo_costs
#Bound 104.10s best:1.15988461e+11 next:[1.15987986e+11,1.15988461e+11] pseudo_costs
#Bound 104.18s best:1.15988461e+11 next:[1.15987986e+11,1.15988461e+11] pseudo_costs
#Bound 106.68s best:1.15988461e+11 next:[1.15988423e+11,1.15988461e+11] pseudo_costs
#Model 106.69s var:1796/16238 constraints:1622/15039 compo:676,292,269,182,67,45,44,42,26,18,...
#Bound 107.66s best:1.15988461e+11 next:[1.15988423e+11,1.15988461e+11] max_lp
#Bound 107.75s best:1.15988461e+11 next:[1.15988424e+11,1.15988461e+11] max_lp
#Bound 107.86s best:1.15988461e+11 next:[1.15988424e+11,1.15988461e+11] max_lp
#Done  107.89s max_lp
#Done  107.89s quick_restart
#Done  107.90s probing

...

[Scaling] scaled_objective_bound: 1.15988e+11 corrected_bound: 1.15987e+11 delta: 1.61722e+06
[Scaling] Warning: OPTIMAL was reported, yet the objective gap (1.10377e+17) is greater than requested absolute limit (0.0001).

CpSolverResponse summary:
status: OPTIMAL
objective: 1.103767340707094e+17
best_bound: 115986843582.3407
booleans: 7336
conflicts: 51165
branches: 92027
propagations: 19941653
integer_propagations: 30257433
restarts: 13641
lp_iterations: 60638
walltime: 147.853
usertime: 2192.12
deterministic_time: 478.549
gap_integral: 3482.85
solution_fingerprint: 0xf2f7f42e4b028458
```
Which is clearly wrong, for the gap/objective value.

Let us debug this, and thanks for the interesting example :-)

**Comment by lperron** [comment]
So, I realized all coefficients are integers, or integer + 0.5. 
I tweaked the model to use the integer objective and I multiplied all coefficients by 2, and solved with a gap limit of 1e-4.

```CpSolverResponse summary:
status: OPTIMAL
objective: 239684166749
best_bound: 239663707682
booleans: 9436
conflicts: 1051
branches: 22443
propagations: 1317260
integer_propagations: 5228750
restarts: 17381
lp_iterations: 37499
walltime: 70.2455
usertime: 951.003
deterministic_time: 133.587
gap_integral: 2321.41
solution_fingerprint: 0xfb9fd569fb693224
```

So, as a workaround, you can scale your coefficients to integers manually, and bypass for floating point computations.

**Comment by martineronneberg** [comment]
Thank you! Is a code change needed to specify that the objective is integer, or is it sufficient to just scale the coefficients and it will be integer automatically?

**Comment by lperron** [comment]
I assume you are using the CP-SAT python directly (and not through
the linear solver).

When calling Minimize() or Maximize(), we scan the objective coefficients.
If all are integral, we create an objective(), otherwise we create a
floating_point_objective().

So in your case, multiply your coefficient by 2 and make sure you cast to
int() when building the objective expression.

Thanks again.
Laurent Perron | Operations Research | ***@***.*** | (33) 1 42 68 53
00
